kubespawner/spawner.py

```diff
@@ -2779,3 +2779,28 @@ async def _ensure_namespace(self):
                 # It's fine if it already exists
                 self.log.exception("Failed to create namespace %s", self.namespace)
                 raise
+
+    async def delete_forever(self):
+        """Called when a user is deleted.
+
+        This can do things like request removal of resources such as persistent storage.
+        Only called on stopped spawners, and is likely the last action ever taken for the user.
+
+        This will only be called once on the user's default Spawner.
+        Supported by JupyterHub 1.4.0+.
+        """
+        try:
+            self.api.read_namespaced_persistent_volume_claim(
+                self.pvc_name, self.namespace
+            )
+        except ApiException as e:
+            if e.status == 404:
+                self.log.warning(
+                    "Could not delete %s. This PVC does not exist.", self.pvc_name
+                )
+            else:
+                raise
+        else:
+            self.api.delete_namespaced_persistent_volume_claim(
+                self.pvc_name, self.namespace
+            )
```

Comment:
@minrk do you have any suggested changes to this section?

I'm a bit uncertain about the use of async without any `await` statement and that we don't have any timeouts defined for example.

For reference, this is the async stop() method:

```python
    async def stop(self, now=False):
        delete_options = client.V1DeleteOptions()

        if now:
            grace_seconds = 0
        else:
            grace_seconds = self.delete_grace_period

        delete_options.grace_period_seconds = grace_seconds

        ref_key = "{}/{}".format(self.namespace, self.pod_name)
        await exponential_backoff(
            partial(
                self._make_delete_pod_request,
                self.pod_name,
                delete_options,
                grace_seconds,
                self.k8s_api_request_timeout,
            ),
            f'Could not delete pod {ref_key}',
            timeout=self.k8s_api_request_retry_timeout,
        )

        try:
            await exponential_backoff(
                lambda: self.pod_reflector.pods.get(ref_key, None) is None,
                'pod %s did not disappear in %s seconds!'
                % (ref_key, self.start_timeout),
                timeout=self.start_timeout,
            )
        except TimeoutError:
            self.log.error(
                "Pod %s did not disappear, restarting pod reflector", ref_key
            )
            self._start_watching_pods(replace=True)
            raise
```

Reply:
> I'm a bit uncertain about the use of async without any await

This in principle is fine, and correct if you are implementing a generally async API but don't have anything to wait for. In this case, however, all k8s API requests should be async via `self.asynchronize`, in which case there will be awaits. Following the timeout/conflict/error handling pattern in our other API requests makes sense.

Yes, I think we do want to make sure to always make kube API calls from the background with retries/error handling, as we do in e.g. `_make_*_request`. We should never call `self.api.anything()` from the main thread.

Reply:
Hi @consideRatio , thanks to you and @minrk for the suggestions! I tried to implement what you outlined above. I tested the changes again with sqlite-pvc and it worked well. Please let me know if any other changes are required.

Reply:
Hi @consideRatio , do the changes look OK?

Reply:
@nsshah1288 thanks for your work on this! :tada:

If the tests pass and the following changes below LGTY I think we could be ready to merge this.

- I cleaned up the merge commit (reset everything to the current master branch and then manually cherry picked all but the merge commit and solving some trivial merge conflicts).
- By doing above I --force pushed to this PRs branch, backing up its previous state in https://github.com/consideRatio/kubespawner/tree/feature/shahn3_pvcDeletion-backup
- I added ea12ea79 because the function didn't have the correct signature matching JupyterHub's Spawner base class
- I opted to not default to delete_grace_period for the PVC deletion request because as it is explicitly described to be used for the pod deletion requests.

  ```python
      delete_grace_period = Integer(
          1,
          config=True,
          help="""
          Time in seconds for the pod to be in `terminating` state before is forcefully killed.
  
          Increase this if you need more time to execute a `preStop` lifecycle hook.
  
          See https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods for
          more information on how pod termination works.
  
          Defaults to `1`.
          """,
      )
  ```

Reply:
Thank you so much for your detailed help on this @consideRatio ! The changes look good to me.